crates/sema/src/ast_passes.rs

```diff
@@ -69,11 +78,32 @@ impl<'ast> Visit<'ast> for AstValidator<'_> {
         }
     }
 
+    fn visit_stmt(&mut self, stmt: &'ast ast::Stmt<'ast>) {
+        let Stmt { kind, .. } = stmt;
+
+        match kind {
+            StmtKind::DoWhile(stmt, ..) => {
+                self.in_loop_depth += 1;
+                self.visit_stmt(stmt);
```

Comment:
these should use `self.walk_stmt` instead of `self.visit_stmt` to recurse on all statements correctly